File: frontend/src/services/productService.ts
import apiClient from './api';

export interface Product {
  productId: number;
  productCode: string;
  productName: string;
  productType?: string;
  specification?: string;
  unit: string;
  standardCycleTime?: number;
  isActive: boolean;
  tenantId: string;
  tenantName: string;
  remarks?: string;
  createdAt: string;
  updatedAt: string;
}

export interface ProductCreateRequest {
  productCode: string;
  productName: string;
  productType?: string;
  specification?: string;
  unit: string;
  standardCycleTime?: number;
  remarks?: string;
}

export interface ProductUpdateRequest {
  productName?: string;
  productType?: string;
  specification?: string;
  unit?: string;
  standardCycleTime?: number;
  remarks?: string;
}

const productService = {
  // Get all products
  getProducts: async (): Promise<Product[]> => {
    const response = await apiClient.get<Product[]>('/products');
    return response.data;
  },

  // Get active products only
  getActiveProducts: async (): Promise<Product[]> => {
    const response = await apiClient.get<Product[]>('/products/active');
    return response.data;
  },

  // Get product by ID
  getProduct: async (id: number): Promise<Product> => {
    const response = await apiClient.get<Product>(`/products/${id}`);
    return response.data;
  },

  // Get product by code
  getProductByCode: async (productCode: string): Promise<Product> => {
    const response = await apiClient.get<Product>(`/products/code/${productCode}`);
    return response.data;
  },

  // Create new product
  createProduct: async (product: ProductCreateRequest): Promise<Product> => {
    const response = await apiClient.post<Product>('/products', product);
    return response.data;
  },

  // Update product
  updateProduct: async (id: number, product: ProductUpdateRequest): Promise<Product> => {
    const response = await apiClient.put<Product>(`/products/${id}`, product);
    return response.data;
  },

  // Delete product
  deleteProduct: async (id: number): Promise<void> => {
    await apiClient.delete<void>(`/products/${id}`);
  },

  // Activate product
  activateProduct: async (id: number): Promise<Product> => {
    const response = await apiClient.post<Product>(`/products/${id}/activate`);
    return response.data;
  },

  // Deactivate product
  deactivateProduct: async (id: number): Promise<Product> => {
    const response = await apiClient.post<Product>(`/products/${id}/deactivate`);
    return response.data;
  },
};

export default productService;

File: backend/src/main/java/kr/co/softice/mes/domain/service/ProductService.java
package kr.co.softice.mes.domain.service;

import kr.co.softice.mes.common.exception.BusinessException;
import kr.co.softice.mes.common.exception.ErrorCode;
import kr.co.softice.mes.domain.entity.ProductEntity;
import kr.co.softice.mes.domain.entity.TenantEntity;
import kr.co.softice.mes.domain.repository.ProductRepository;
import kr.co.softice.mes.domain.repository.TenantRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Product Service
 * 제품 마스터 서비스
 *
 * @author Moon Myung-seop
 */
@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;
    private final TenantRepository tenantRepository;

    /**
     * Find all products by tenant ID
     */
    public List<ProductEntity> findByTenant(String tenantId) {
        return productRepository.findByTenantIdWithTenant(tenantId);
    }

    /**
     * Find active products by tenant ID
     */
    public List<ProductEntity> findActiveByTenant(String tenantId) {
        return productRepository.findByTenantIdAndIsActiveWithTenant(tenantId, true);
    }

    /**
     * Find product by ID
     */
    public Optional<ProductEntity> findById(Long productId) {
        return productRepository.findById(productId);
    }

    /**
     * Find product by product code
     */
    public Optional<ProductEntity> findByProductCode(String tenantId, String productCode) {
        return productRepository.findByTenant_TenantIdAndProductCode(tenantId, productCode);
    }

    /**
     * Create new product
     */
    @Transactional
    public ProductEntity createProduct(ProductEntity product) {
        log.info("Creating product: {} for tenant: {}",
            product.getProductCode(), product.getTenant().getTenantId());

        // Check duplicate
        if (productRepository.existsByTenantAndProductCode(product.getTenant(), product.getProductCode())) {
            throw new BusinessException(ErrorCode.PRODUCT_ALREADY_EXISTS);
        }

        return productRepository.save(product);
    }

    /**
     * Update product
     */
    @Transactional
    public ProductEntity updateProduct(ProductEntity product) {
        log.info("Updating product: {}", product.getProductId());

        if (!productRepository.existsById(product.getProductId())) {
            throw new BusinessException(ErrorCode.PRODUCT_NOT_FOUND);
        }

        return productRepository.save(product);
    }

    /**
     * Delete product
     */
    @Transactional
    public void deleteProduct(Long productId) {
        log.info("Deleting product: {}", productId);
        productRepository.deleteById(productId);
    }

    /**
     * Activate product
     */
    @Transactional
    public ProductEntity activateProduct(Long productId) {
        ProductEntity product = productRepository.findById(productId)
            .orElseThrow(() -> new BusinessException(ErrorCode.PRODUCT_NOT_FOUND));

        product.setIsActive(true);
        return productRepository.save(product);
    }

    /**
     * Deactivate product
     */
    @Transactional
    public ProductEntity deactivateProduct(Long productId) {
        ProductEntity product = productRepository.findById(productId)
            .orElseThrow(() -> new BusinessException(ErrorCode.PRODUCT_NOT_FOUND));

        product.setIsActive(false);
        return productRepository.save(product);
    }
}
